Upgrade webpack to version 5

thealexandrelara opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
The current version of Webpack in this project is version 4. The goal is to upgrade to Webpack 5 to get improved build performance, access to new features and improvements, improved security, and compatibility with JavaScript. Once the upgrade is complete, and we've thoroughly tested the changes to our build system to ensure they don't cause any regressions or break existing functionality, we can enable HRM.

<!-- Reference any related issues or PRs here -->

Fixes #5244 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing instructions

#### Scenario 1: Build
1. Run `npm run build` and make sure the build finishes correctly;
2. Run `npm start` and check if the application is running without any error;

#### Scenario 2: Blocks
1. Log in to your WordPress dashboard.
2. Add a new post/page;
3. Add multiple blocks to the post/page;
4. Make sure you can edit the blocks settings and styles;
5. Click on the Save button;
6. Visit the post/page on the frontend and make sure everything is working.

#### Scenario 3: Templates
1. Log in to your WordPress dashboard.
2. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
3. On the left-hand side menu, click on Appearance > Editor > Templates
4. Select and click on a template from the list to start editing it. If it is a Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before;
5. Make sure everything is working.
6. Add multiple blocks to the template;
7. Make sure you can edit the blocks settings and styles;
8. Click on the Save button;
9. Visit the template page on the frontend and make sure everything is working.
10. Repeat steps 4-9 to test all the available templates.

#### Scenario 4: Testing with generated zip file in a hosted website
1. Run `npm run package-plugin:deploy`. This command will generate a zip file in the root folder of your project;
2. Create a new WooCommerce on [Pressable](https://pressable.com/) or any other hosting of your preference;
3. From your WordPress dashboard, go to Plugins > Add new. Upload the generated zip and activate the plugin;
4. Repeat the steps from the scenarios 2 and 3 above.

#### Scenario 5: Upgrade from previous version
> **_ Recommended pre-requisite: A hosted website with a version of the plugin using Webpack 4_**
1. Run `npm run package-plugin:deploy`. This command will generate a zip file in the root folder of your project;
2. Repeat the steps from the scenarios 2 and 3, above.
3. Considering that you have a site with a previous version of the plugin installed (a version that still uses webpack 4). Go to Plugins > Installed plugins. Disable the current plugin version.
4. Upload the generated zip and activate the plugin;
5. Make sure that the blocks and templates are working both on the editor and on the frontend.

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Upgrade Webpack from version 4 to version 5

thealexandrelara committed:
Upgrade terser-webpack-plugin to version 4.2.3

thealexandrelara committed:
Upgrade webpack-bundle-analyzer to 4.7.0

thealexandrelara committed:
Upgrade to Webpack version 5

thealexandrelara committed:
Upgrade @wordpress/dependency-extraction-webpack-plugin to 4.6.0

thealexandrelara committed:
Upgrade dependency copy-webpack-plugin to version 11.0.0

thealexandrelara committed:
Upgrade dependency terser-webpack-plugin to version 5.3.6

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Replace webpack-rtl-plugin with the new @automattic/webpack-rtl-plugin

thealexandrelara committed:
Replace module.issuer with the new ModuleGraph API

There is a warning appearing in the console when running the application. This is due to the fact that the module.issuer has been deprecated on Webpack 5 and replaced with the new ModuleGraph API. This commit replaces the deprecated API with the new one.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Upgrade babel and babel plugins to latest version

thealexandrelara committed:
Replace jsonpFunction with the new uniqueName property

Add a unique name of the webpack build to avoid multiple webpack runtimes to conflict when using globals. It defaults to output.library name or the package name from package.json in the context, if both aren't found, it is set to an ''.

thealexandrelara committed:
Replace cacheDirectory inline configuration with options.cacheDirectory

thealexandrelara committed:
Upgrade @wordpress/e2e-tests dependency to version 5.6.0

thealexandrelara committed:
Remove babel-plugin-transform-react-jsx dependency

Remove babel-plugin-transform-react-jsx dependency because it is already included in @wordpress/babel-preset-default

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Remove unnecessary Babel dependencies

Remove some unnecessary babel dependencies that are already included in the @babel/preset-env package.

thealexandrelara committed:
Upgrade puppeteer dependency to version 16.2.0

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

gigitux commented:
Thanks for working on this! 
Do you think that it is fine having `${number}-frontend`? I remember that I tried to fix it without success.
On the PHP side, we register scripts with the structure `wc-${block_name}-block-frontend`.

![image](https://user-images.githubusercontent.com/4463174/209932615-5e6462c4-2bc3-4aff-b08e-2771d0a2ba39.png)

Aljullu commented on package.json:
```diff
@@ -91,12 +91,13 @@
-		"@babel/plugin-syntax-jsx": "7.18.6",
-		"@babel/polyfill": "7.12.1",
-		"@babel/preset-typescript": "7.18.6",
+		"@automattic/webpack-rtl-plugin": "^6.0.0",
+		"@babel/cli": "^7.20.7",
+		"@babel/core": "^7.20.7",
+		"@babel/plugin-proposal-class-properties": "^7.18.6",
+		"@babel/plugin-syntax-jsx": "^7.18.6",
+		"@babel/polyfill": "^7.12.1",
+		"@babel/preset-typescript": "^7.18.6",
```
If I remember correctly, we use fixed versions in `package.json` (_aka_, without the `^` caret operator), because renovatebot already takes care of updating the dependencies.

thealexandrelara replied:
> Really exciting to see the bundle size improvements webpack 5 brings! 🥳

Yeahh, that's really awesome to see the bundle size decreasing that much 🎉

> I did a partial review so far, and it looks promising. However, I noticed Storybook is broken (`npm run storybook`), even though in `trunk` it worked fine. Can you reproduce as well?

Thank you for pointing this out! I just fixed that. By default, Storybook uses Webpack 4 to build everything, [so I tweaked its configurations](https://storybook.js.org/docs/react/builders/webpack#webpack-5) to make it use Webpack 5


> If I remember correctly, we use fixed versions in `package.json` (_aka_, without the `^` caret operator), because renovatebot already takes care of updating the dependencies.

Great, I just removed the caret operator from the versions there. Thank you!

Aljullu reviewed (commented):
Really exciting to see the bundle size improvements webpack 5 brings! :partying_face: 

I did a partial review so far, and it looks promising. However, I noticed Storybook is broken (`npm run storybook`), even though in `trunk` it worked fine. Can you reproduce as well?

thealexandrelara committed:
Remove caret from package.json dependencies

thealexandrelara committed:
Fix Storybook build error

This commit fixes the Storybook build error that was being caused because of Storybook by default uses Webpack 4, but since we are currently upgrading our webpack to version 5, we need to install some required dependencies and also explicitly tell Storybook to use Webpack 5.

thealexandrelara commented:
> Thanks for working on this! Do you think that it is fine having `${number}-frontend`? I remember that I tried to fix it without success. On the PHP side, we register scripts with the structure `wc-${block_name}-block-frontend`.
> 
> ![image](https://user-images.githubusercontent.com/4463174/209932615-5e6462c4-2bc3-4aff-b08e-2771d0a2ba39.png)

Oh nice catch! You mean because some numeric files were created during the build? I would have to investigate why Webpack made those files after the upgrade. Can you share from which file you took this code (in your image)?

gigitux commented:
> Oh nice catch! You mean because some numeric files were created during the build? 

Yes! 

> I would have to investigate why Webpack made those files after the upgrade. 

👍 . Otherwise, we can check that this new filename doesn't impact the project. 

> Can you share from which file you took this code (in your image)?

There are multiple instances, search in the project with the keyword "-block-frontend" and you will find multiple results.

I took the screenshot from `Cart.php`.

thealexandrelara commented:
> > Oh nice catch! You mean because some numeric files were created during the build?
> 
> Yes!
> 
> > I would have to investigate why Webpack made those files after the upgrade.
> 
> 👍 . Otherwise, we can check that this new filename doesn't impact the project.
> 
> > Can you share from which file you took this code (in your image)?
> 
> There are multiple instances, search in the project with the keyword "-block-frontend" and you will find multiple results.
> 
> I took the screenshot from `Cart.php`.

I've finally discovered what is actually happening. By default, Webpack 5 splits modules that are shared between multiple entries. I've found that the shared chunks were already happening before; however, chunk filenames were different because they were using the `splitChunks.automaticNameDelimiter` property in the Webpack config to name chunks. So, for example, if we had Cart and Checkout components sharing the same piece of code, it would result in a file `cart--checkout-frontend.js`.

When it comes to Webpack 5 when they generate shared chunks, [those chunks do not have a name associated with them](https://github.com/webpack/webpack/discussions/14451), but they have an id (that is numeric), so they use that `id` to name the file. Looking at the code, it seems the PHP files are just importing the entry files (i.e.: `cart-frontend.js`, `checkout-frontend.js`), and Webpack is responsible for looking at the chunk graph and importing the required chunks for that piece of code to work. It was the same before, but with a different naming schema, so I think this wouldn't have a negative impact on the project. We can keep testing and check if any weird behavior appears.

Aljullu reviewed (commented):
Confirming Storybook works well now. I also did some smoke testing and most things work as expected, however, I noticed that on RTL languages, styles seem to be broken.

This is with the language set to Arabic:

`trunk` | This branch
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/211557831-15e7ca65-f348-42fd-b220-ca8b1147203a.png) | ![imatge](https://user-images.githubusercontent.com/3616980/211557893-9c2a6f17-d4e8-44e8-b4e3-b3226bdc238e.png)

Can you reproduce as well, @thealexandrelara?

kmanijak commented:
I was checking the security vulnerability alerts and I think [this one](https://github.com/woocommerce/woocommerce-blocks/security/dependabot/16) will be possible fixed after updating Webpack to version 5.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix package-lock.json after merging with trunk

thealexandrelara committed:
Add own webpack-rtl-plugin implementation to the project

Before upgrading Webpack to version 5, we were using the original `webpack-rtl-plugin` released by Romain Berger; unfortunately, this plugin is not compatible with Webpack 5, so we replaced it with `@automattic/webpack-rtl-plugin`. The problem is that `@automattic/webpack-rtl-plugin` by default generates files with the '.rtl.css' suffix and does not provide a way to change that.

This commit adds our own implementation of the `webpack-rtl-plugin` (adapted from `@automattic/webpack-rtl-plugin`) that is compatible with Webpack 5 and allows us to change the suffix of the generated files to follow the recommended way defined by Wordpress (https://codex.wordpress.org/Right-to-Left_Language_Support)

thealexandrelara commented:
> Confirming Storybook works well now. I also did some smoke testing and most things work as expected, however, I noticed that on RTL languages, styles seem to be broken.
>
> Can you reproduce as well, @thealexandrelara?

I can't reproduce it, but I made some changes to the Webpack config to generate css files with the suffix `-rtl.css` as [recommended by WordPress](https://codex.wordpress.org/Right-to-Left_Language_Support). The library that we are using (@automattic/webpack-rtl-plugin) generates files with the suffix `.rtl.css` and, unfortunately, does not provide an option to change that

Aljullu commented on bin/webpack-configs.js:
```diff
@@ -643,13 +623,22 @@ const getStylingConfig = ( options = {} ) => {
-									/[\\/]assets[\\/]js[\\/]editor-components[\\/]/
-								) ),
+						test: ( module = {}, { moduleGraph } ) => {
+							if ( ! module.type.includes( 'css' ) ) return false;
```
Nit: I think it's better to always use multiline clauses. This will keep consistency with our codebase and IMO makes code easier to read.
```suggestion
							if ( ! module.type.includes( 'css' ) ) {
								return false;
							}
```

Aljullu reviewed (commented):
Thanks for updating this PR, @thealexandrelara! This fixes the issue for me; however, I'm not convinced of having to maintain a copy of `webpack-rtl-plugin`. IMO it would be better to:

1. Try to upstream to `@automattic/webpack-rtl-plugin` the option to use dashes instead of dots (`-rtl.css` instead of `.rtl.css`).
2. Use `rtlcss` directly, in a similar fashion to [what Gutenberg is doing](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L128-L134).

What do you think? I'm a bit wary that having a copy of `webpack-rtl-plugin` will increase the amount of maintenance work we have to do in the future as we need to keep up with updates.

thealexandrelara committed:
Change conditional clause to be multiline

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix package-lock.json after merge with trunk

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix package-lock.json after merge with trunk

thealexandrelara commented:
> Thanks for updating this PR, @thealexandrelara! This fixes the issue for me; however, I'm not convinced of having to maintain a copy of `webpack-rtl-plugin`. IMO it would be better to:
> 
> 1. Try to upstream to `@automattic/webpack-rtl-plugin` the option to use dashes instead of dots (`-rtl.css` instead of `.rtl.css`).
> 2. Use `rtlcss` directly, in a similar fashion to [what Gutenberg is doing](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L128-L134).
> 
> What do you think? I'm a bit wary that having a copy of `webpack-rtl-plugin` will increase the amount of maintenance work we have to do in the future as we need to keep up with updates.

Thank you for reviewing and for the suggestions! I think the first option would be preferable because some things concern me about the second option:
1. To make it work, Gutenberg is making use of Workers to manipulate the files during the build. They have to manually [render the SASS files](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L115), process the result  using postcss, and also manually manipulate the files ([retrieving](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L84), [changing the filenames ](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L85)and [saving them](https://github.com/WordPress/gutenberg/blob/d5915916abc45e6682f4bdb70888aa41e98aa395/bin/packages/build-worker.js#L137));
2. They also make use of a `stream` object that stores and processes the files, including being responsible for building the css;
3. They are using the `node-worker-farm` lib to create the worker responsible for processing the css, but that library is not being maintained anymore;

It seems most of the code is from 2-4 years ago, but since then, Webpack has evolved, and we can probably benefit from some new built-in resources without having to manually manipulate and process the files. 

Considering that, I think the best option would be to upstream the changes to `@automattic/webpack-rtl-plugin`. One thing we should consider is if we have to block this card until they release a new version that allows us to change the filename extension or if we can use this temporary solution (with our version of /the Webpack RTL Plugin) and open a new issue to replace this solution with the latest version in the future. What do you think?

Aljullu reviewed (approved):
> Considering that, I think the best option would be to upstream the changes to `@automattic/webpack-rtl-plugin`. One thing we should consider is if we have to block this card until they release a new version that allows us to change the filename extension or if we can use this temporary solution (with our version of /the Webpack RTL Plugin) and open a new issue to replace this solution with the latest version in the future. What do you think?

That's a good point. Probably no need to block updating to Webpack 5 because of this. Will you open an issue about this, though? This way we don't forget. :slightly_smiling_face: 

I did another round of testing and everything is looking great on my end. Good work, @thealexandrelara! :ship:

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara commented:
> Will you open an issue about this, though? This way we don't forget. 🙂
> 
> I did another round of testing and everything is looking great on my end. Good work, @thealexandrelara! 🚢

Sure I can open an issue about this! Thank you 🙌

thealexandrelara committed:
Rename files to fix ESLint errors

This commit renames files that have the .js extension but contain JSX code. This is causing ESLint to throw errors because by default our Eslint configuration expects only files with the .jsx extension to contain JSX code.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara commented:
Just providing a quick update on why this PR was not merged yet: I was reviewing the Github checks and noticed that some unit tests are broken on this branch, probably caused by a babel configuration. I'm currently working on other issues that were prioritized, but I'll be working on this one again as soon as I can

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix package-lock.json file

thealexandrelara committed:
Add is-plain-obj module to the transformIgnorePatterns of jest config

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

ralucaStan commented:
@thealexandrelara I am converting this to draft. Feel free to remove this once this PR is opened for review

opr commented:
Removed the milestone - please re-add it when this PR gets merged.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Update package-lock.json

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix package-lock.json

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Upgrade @wordpress/i18n dependency to version 4.31.0

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Update package-lock.json

thealexandrelara committed:
Update composer lock file

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix Webpack config for Webpack 5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Add the package-lock.json

thealexandrelara committed:
Remove unsupported config from webpack

thealexandrelara committed:
Fix error with Webpack build

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Add wait for network idle to the tests

thealexandrelara committed:
Attempt to fix e2e test

thealexandrelara committed:
Restore promise.all

thealexandrelara committed:
Upgrade puppeteer to v17.1.3

thealexandrelara committed:
Upgrade expect-puppeteer

thealexandrelara committed:
Update expect-puppeteer

thealexandrelara committed:
Downgrade expect-puppeteer

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Revert "Upgrade puppeteer to v17.1.3"

This reverts commit f246e256cd0ce26bd378f0f39676b51030499445.

thealexandrelara committed:
Upgrade Puppeteer to version 17.1.3

thealexandrelara committed:
Fix executionContext.frame is not a function error

thealexandrelara committed:
Fix e2e tests

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Remove isExperimentalBuild from Product Gallery inner blocks

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Upgrade Webpack and Webpack-cli to latest version

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Upgrade postcss and mini-css plugins

thealexandrelara committed:
Fix error with mini-cart block

nefeline reviewed (commented):
Hey @thealexandrelara ! I'm noticing some oddities in the editor when inserting certain blocks or patterns. 

For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

<img width="700" alt="Screenshot 2023-08-10 at 09 22 41" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/6d4fbeb2-9c5f-428a-a8e7-84361bb201cf">

Are you able to reproduce this as well?

tjcafferkey commented:
>For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

@nefeline I ran a clean install and build and I'm not encountering this problem. Just wondering whether you've done the same?

Below is a snippet of my zsh command, and a screenshot of the filter blocks in editor.

```
function wcb-fresh-install {
  cd ~/LocalSites/woodevelopment/app/public/wp-content/plugins/woocommerce-blocks
  nvm-install
  rm -rf node_modules
  composer install
  npm install
  npm run build
  npm run start
}
```

![Screenshot 2023-08-10 at 08 56 36](https://github.com/woocommerce/woocommerce-blocks/assets/8639742/9e56c1fb-c620-4be1-b072-9bb993f87047)

nefeline commented:
> I ran a clean install and build and I'm not encountering this problem. Just wondering whether you've done the same?

I ran the following:

- nvm use
- rm -rf node_modules
- npm install
- npm run build

Composer should not be interfering with this webpack update, but I also ran it on a second test just in case, I'm getting the same results:

![Screen Recording 2023-08-10 at 11 11 33](https://github.com/woocommerce/woocommerce-blocks/assets/15730971/3e612ee7-c405-4d5c-86e3-8e570344afd7)

thealexandrelara committed:
Fix styling error with filter blocks

thealexandrelara commented:
> Hey @thealexandrelara ! I'm noticing some oddities in the editor when inserting certain blocks or patterns.
> 
> For example, the filter by attribute, filter by stock status, and the filter by rating blocks have all display issues in the editor:

> Are you able to reproduce this as well?

Hey @nefeline, thank you for reporting this, I was able to reproduce this as well and just pushed a fix to it. Let me know if this solves the issue for you too

nefeline reviewed (approved):
> Hey @nefeline, thank you for reporting this, I was able to reproduce this as well and just pushed a fix to it. Let me know if this solves the issue for you too

Thanks for the update @thealexandrelara! I can confirm https://github.com/woocommerce/woocommerce-blocks/pull/8013/commits/426670e9d5cdf2069904e3ad8b25ee130b11020f solves the problem.

I ran a series of tests with the following blocks:

* All filter blocks: filter by stock, filter by attribute, filter by rating, filter by price
* Active Filters
* Products (Beta)
* Single Product
* Products by Category
* Products by Attribute
* Mini-cart
* Product Search
* Product tags
* Featured Product

They all work as expected, and the styles are correct when running both on development and production builds. 

Can we ensure this branch is up-to-date with trunk and solve the conflicts, please? I'll do another pass & re-review as soon as everything is up-to-date.

I'm pre-approving as the styles are all correct, and blocks are fully functional now: but we should wait for others to finish their tests as planned on pdToLP-Ll-p2 before merging.

Great work! 🎉

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

nefeline commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/index.tsx:
```diff
@@ -12,11 +11,9 @@ import { Edit, Save } from './edit';
 import metadata from './block.json';
 import './style.scss';
 
-if ( isExperimentalBuild() ) {
```
Should we ditch these changes? Any particular reason for us to be removing `isExperimentalBuild` here?

thealexandrelara replied:
The Product Gallery block contains some inner blocks. These inner blocks are only available inside the Product Gallery block, so even if I remove this, they will only be made available on the editor or frontend with its main block. 

The reason why I removed this is because of an issue that was occurring when using the `isExperimentalBuild()` (p1691416552602699/1691413062.084399-slack-C02UBB1EPEF) inside the inner blocks. I don't know exactly why this was happening, but I suspect the cause is that for the inner blocks to work, we have to import them in the `index.tsx` file of the [Product Gallery block](https://github.com/woocommerce/woocommerce-blocks/blob/chore/upgrade-webpack-to-version-5/assets/js/blocks/product-gallery/index.tsx#L14-L15), since there are many changes/improvements on the tree-shaken process in Webpack V5, sometimes Webpack can fail to recognize a file/piece of code that should be included in the bundle, in this particular case, it failed to include the `feature_flags` function that the `isExperimentalBuild()` call uses.

nefeline replied:
> The Product Gallery block contains some inner blocks. These inner blocks are only available inside the Product Gallery block, so even if I remove this, they will only be made available on the editor or frontend with its main block.

Gotcha, thanks for clarifying.

> The reason why I removed this is because of an issue that was occurring when using the isExperimentalBuild() (p1691416552602699/1691413062.084399-slack-C02UBB1EPEF) inside the inner blocks. I don't know exactly why this was happening, but I suspect the cause is that for the inner blocks to work, we have to import them in the index.tsx file of the [Product Gallery block](https://github.com/woocommerce/woocommerce-blocks/blob/chore/upgrade-webpack-to-version-5/assets/js/blocks/product-gallery/index.tsx#L14-L15), since there are many changes/improvements on the tree-shaken process in Webpack V5, sometimes Webpack can fail to recognize a file/piece of code that should be included in the bundle, in this particular case, it failed to include the feature_flags function that the isExperimentalBuild() call uses.

👍 thanks for the detailed explanation! If this only affects inner blocks and won't cause any problems for the `product-gallery` or the `product-collection` blocks (these are the ones currently behind the experimental flag), all good.

nefeline commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/index.tsx:
```diff
@@ -12,13 +11,11 @@ import { Edit } from './edit';
 import metadata from './block.json';
 import './style.scss';
 
-if ( isExperimentalBuild() ) {
```
Same as the previous: should we move this to a separate PR?

thealexandrelara replied:
We can try solving this in a separate PR if we consider it necessary. The only thing we need to have in mind is that this is an inner block of the Product Gallery block that is currently set to be experimental, which means that all its inner blocks will only be made available when used inside the Product Gallery block and only if this block is also available.

nefeline replied:
> We can try solving this in a separate PR if we consider it necessary. The only thing we need to have in mind is that this is an inner block of the Product Gallery block that is currently set to be experimental, which means that all its inner blocks will only be made available when used inside the Product Gallery block and only if this block is also available.

All good, thanks for the explanation on https://github.com/woocommerce/woocommerce-blocks/pull/8013#discussion_r1291708015 : I agree it makes sense to keep this change (but only if this limitation with the `isExperimentalBuild()` is exclusively affecting inner blocks and won't expose the gallery and product collection blocks to production ahead of time).

nefeline replied:
BTW: do we have an issue opened to keep track of this limitation with the `isExperimentalBuild()` on inner blocks?

thealexandrelara replied:
As of now, we don't have an issue opened specifically for this, but it's something worth considering. What would be expected from an issue regarding this limitation? Would the focus be on documenting this as a limitation, advising developers against utilizing isExperimentalBuild for inner blocks when the parent block is already using it? Or, would the idea be to find a way of enabling the use of multiple isExperimentalBuild instances in the parent block and its inner blocks?

nefeline replied:
> What would be expected from an issue regarding this limitation?

IMO it should work independent of the block type:  `isExperimentalBuild()` should work as expected for both blocks and inner blocks (as it used to work before this Webpack update).

thealexandrelara replied:
@nefeline After adding the `"./assets/js/settings/blocks/index.ts"` file to the `sideEffects` property of the `package.json` file, the issue is no longer happening, so I restored these files to how it was before 🙌

thealexandrelara committed:
Fix issue when running unit tests

nefeline commented:
@thealexandrelara I ran a second round of tests here and confirmed blocks are working as expected & styles are correct: thanks again for your prompt response & work here! 🙌! Let's see if anyone has any other findings based on the remaining tests: pdToLP-Ll-p2

gigitux reviewed (approved):
I verified that E2E test env setup works correctly locally! LGTM!

opr reviewed (approved):
As mentioned in pdToLP-Ll-p2#comment-542 I have tested the Cart and Checkout blocks and could not find any issues!

alexflorisca commented:
I've ran the build and start commands with the new webpack config and all seems fine for me too. The only thing I wanted to point out is that several of the bundles scripts have increased significantly in size. See the  bundle size https://github.com/woocommerce/woocommerce-blocks/pull/8013#issuecomment-1362082382 above - look for any with a 🆘 or 🚨 emojis. Have we addressed this?

I also did a quick test on the Cart page to see what the overall increase in assets is

| Cart Page 	| Webpack 4 	| Webpack 5 	|
|-----------	|-----------	|-----------	|
| JS        	| 4.1MB     	| 4.1MB     	|
| CSS       	| 335Kb     	| 351Kb     	|

The JS looks fine overall but the CSS has increased slightly and I wonder why that is? I also wonder if we should do this for other blocks, or maybe one page with all the blocks on it to see if there's any significant impact?

Think this is quite important to address before we merge as we don't want the bundle size to increase

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

dinhtungdu commented:
@thealexandrelara I got this error while testing storybook against this PR:

```sh
**99% done plugins webpack-hot-middleware**webpack built preview 078da4b21ec586873670 in 1732ms
```

The script stucks at 99%. Viewing the storybook on the frontend, I only see the spinner running forever

<img width="1720" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/5423135/88df4870-3892-4d39-83a9-610d518ee4cf">


I see you attempted to fix the storybook in https://github.com/woocommerce/woocommerce-blocks/tree/a82c20da5881d2b957f6a609e532b7a6d5a3cb51, I wonder if it's still working for you now.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara commented:
> @thealexandrelara I got this error while testing storybook against this PR:
> 
> ```shell
> **99% done plugins webpack-hot-middleware**webpack built preview 078da4b21ec586873670 in 1732ms
> ```
> 
> The script stucks at 99%. Viewing the storybook on the frontend, I only see the spinner running forever
> 
> <img alt="image" width="1720" src="https://user-images.githubusercontent.com/5423135/261252862-88df4870-3892-4d39-83a9-610d518ee4cf.png">
> I see you attempted to fix the storybook in https://github.com/woocommerce/woocommerce-blocks/tree/a82c20da5881d2b957f6a609e532b7a6d5a3cb51, I wonder if it's still working for you now.

You're right, I just tested it and can reproduce the issue as well. I attempted to resolve it by cleaning the Storybook cache by running `rm -rf node_modules/.cache/storybook`, but unfortunately, it didn't fix the issue.

I'll need to investigate what's causing this problem, given that it was working before. Perhaps upgrading to the latest version is the only option, although I was hoping to handle upgrades of external libraries through separate PRs.

dinhtungdu commented:
@thealexandrelara I [tried](https://github.com/woocommerce/woocommerce-blocks/commit/b321aee83a280d5b602f4c2b39dadbeea3079cf5) using `storyStorev7` as Gutenberg is using but got new issues with some stories. I also tried [upgrading to Storybook v7](https://github.com/woocommerce/woocommerce-blocks/tree/try/storybook7) but got the same issues. I don't know how to fix those issues, so posting my trials here hoping it can help you or other fix them.

<img width="1435" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/5423135/59683926-be0b-4a6c-b96d-5e3826f5002d">

kmanijak reviewed (approved):
I checked the blocks if they render correctly and if they're functional.

For each of the blocks in [this list](https://wordpress.org/plugins/woo-gutenberg-products-block/) + Products (Beta) block:
-  ✅ add the block to the post before the upgrade. Upgrade WC Blocks to the version built with Webpack 5. No problems were found either in Editor or Frontend
-  ✅ add the block to the new post after the upgrade. No problems were found either in Editor or Frontend

nielslange reviewed (approved):
I've successfully tested the translations against the Webpack 5 update. There was only one minor translation issue, which will solve itself once this PR has been deployed to *.org and the GlotPress translations have been updated, see p1692365786355899/1692363111.353269-slack-C02UBB1EPEF.

imanish003 reviewed (approved):
Hello @thealexandrelara, great job! I've tested the Mini Cart block, and I'm pleased to report that it's functioning as anticipated. 🚀

samueljseay commented:
@thealexandrelara @dinhtungdu I had a go at fixing the storybook issues today.

You can fix the immediate problem of unexpected token default,its a docgen issue, it doesn't like static class properties and mangles them.

Install `react-docgen-typescript-plugin` then add this to the main.js:

```
typescript: {
		reactDocgen: 'react-docgen-typescript-plugin',
},
```

(note that upgrading to storybook 7 I think also fixes this problem due to them updating the dep on docgen plugin, but if you wanted a way to avoid upgrading to storybook 7 then this is an option 😄 )

That does not however solve the `exports is not defined` error. That's coming from `rungen` which is a dep of `@wordpress/redux-routine`. I haven't been able to figure this one out yet, and also worth nothing that although storybook appears to display more correctly on `trunk`, it also, suffers from this exception so Storybook is already a bit broken.

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Fix storybook script not loading

thealexandrelara committed:
Fix a11y issue in Storybook

thealexandrelara commented:
Hey everyone,
Thank you for testing and approving the PR. Really great to see that we are almost there and hopefully soon we will be able to merge this PR 🙌 

> @thealexandrelara @dinhtungdu I had a go at fixing the storybook issues today.
> 
> You can fix the immediate problem of unexpected token default,its a docgen issue, it doesn't like static class properties and mangles them.
> 
> Install `react-docgen-typescript-plugin` then add this to the main.js:
> 
> ```
> typescript: {
> 		reactDocgen: 'react-docgen-typescript-plugin',
> },
> ```
@samueljseay Ohh thank you so much!!! I just pushed the changes as you suggested and the Storybook is working again!


> 
> (note that upgrading to storybook 7 I think also fixes this problem due to them updating the dep on docgen plugin, but if you wanted a way to avoid upgrading to storybook 7 then this is an option 😄 )
> 
> That does not however solve the `exports is not defined` error. That's coming from `rungen` which is a dep of `@wordpress/redux-routine`. I haven't been able to figure this one out yet, and also worth nothing that although storybook appears to display more correctly on `trunk`, it also, suffers from this exception so Storybook is already a bit broken.

Thank you for investigating this, in this case, I believe we should address this problem in a separate PR, probably we can create an issue to upgrade Storybook and work on a fix to this since it is already broken on `trunk`

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

wavvves reviewed (approved):
I tested build, and cart/checkout templates local built branch and via zip, and spotted no different behaviour.

thealexandrelara committed:
Fix error when multiple isExperimentalBuild was being used

dinhtungdu reviewed (approved):
Storybook is testing great with the latest commit. 🎉 Thank you so much for the great work @thealexandrelara and @samueljseay!

`ProductPrice` store has an error but the current `trunk` storybook has that same error too, so it shouldn't block this PR.
<img width="1752" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/5423135/40f3eb42-5437-4744-b5a7-e24e66d2519c">

thealexandrelara committed:
Prevent error when layout is not present in the attributes object

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Update `chunkIds` to `named` in Webpack

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Add cache groups to the Webpack configs

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

tarhi-saad commented:
@thealexandrelara, I bumped this PR to the next release!

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5

thealexandrelara committed:
Merge branch 'trunk' into chore/upgrade-webpack-to-version-5
